myst_nb/render_outputs.py

```diff
@@ -91,6 +92,58 @@ def load_renderer(name: str) -> "CellOutputRendererBase":
     raise MystNbEntryPointError(f"No Entry Point found for myst_nb.mime_render:{name}")
 
 
+RGX_CARRIAGERETURN = re.compile(r".*\r(?=[^\n])")
+RGX_BACKSPACE = re.compile(r"[^\n]\b")
+
+
+def coalesce_streams(outputs: List[NotebookNode]) -> List[NotebookNode]:
+    """Merge all stream outputs with shared names into single streams.
+
+    This ensure deterministic outputs.
+
+    Adapted from:
+    https://github.com/computationalmodelling/nbval/blob/master/nbval/plugin.py.
+    """
+    if not outputs:
+        return []
+
+    new_outputs = []
+    streams = {}
+    for output in outputs:
+        if output["output_type"] == "stream":
+            if output["name"] in streams:
+                streams[output["name"]]["text"] += output["text"]
+            else:
+                new_outputs.append(output)
+                streams[output["name"]] = output
+        else:
+            new_outputs.append(output)
+
+    # process \r and \b characters
+    for output in streams.values():
+        old = output["text"]
+        while len(output["text"]) < len(old):
+            old = output["text"]
+            # Cancel out anything-but-newline followed by backspace
+            output["text"] = RGX_BACKSPACE.sub("", output["text"])
+        # Replace all carriage returns not followed by newline
+        output["text"] = RGX_CARRIAGERETURN.sub("", output["text"])
+
+    # We also want to ensure stdout and stderr are always in the same consecutive order,
+    # because they are asynchronous, so order isn't guaranteed.
+    for i, output in enumerate(new_outputs):
+        if output["output_type"] == "stream" and output["name"] == "stderr":
+            if (
+                len(new_outputs) >= i + 2
+                and new_outputs[i + 1]["output_type"] == "stream"
+                and new_outputs[i + 1]["name"] == "stdout"
+            ):
+                stdout = new_outputs.pop(i + 1)
+                new_outputs.insert(i, stdout)
```

Comment:
So I understand correctly, is this correct:

`new_outputs` may contain a bunch of outputs, some `stdout` and some `stderr`, all interwoven. This section sorts them so that the `stdout` outputs always come after the `stderr` outputs?

and if that is correct, maybe I'm missing something but could we simplify with something like:

```python
sorted(new_outputs, lambda output: output.get("name"))
```

?

Reply:
`stdout` come before `stderr`. Again, I would emphasise this is copied directly from https://github.com/computationalmodelling/nbval/blob/master/nbval/plugin.py. If you want to change this, I would suggest making a PR there first

Reply:
@choldgraf I think it's not what you describe (although I might have misread it).

- non-stream outputs aren't reshuffled. So `[stdout, image, stdout]` stays untouched.
- multiple arguments of the same stream type are merged.

> Again, I would emphasise this is copied directly from https://github.com/computationalmodelling/nbval/blob/master/nbval/plugin.py. If you want to change this, I would suggest making a PR there first

Looking at the description of nbval, I think its purpose is different from publishing notebooks, and therefore the logic it applies might not be optimal for myst-nb or vice versa.

Speaking of publication purposes, reordering streams may be undesirable: imagine demonstrating code that produces a warning halfway through.

On the other hand, coalescing same stream types may be a good idea without even giving the users an option to configure this. Or is there a use case for having two independent text outputs in a row after a single input?

Reply:
>  imagine demonstrating code that produces a warning halfway through.

except, if you write to `sderr`, then there is no guarantee that it will end up halfway `stdout`


anyhow, I have no capacity left to devote to this, so I'll leave you guys to propose new PRs

Reply:
Good point @akhmerov - I feel like we should see how folks use this feature and see if there are any natural points to improve the ux over time 👍